[FIX] hr_expense: Harmonize states synchronization

JulienAlardot opened the pull request:
See task for details but in resume, for company-paid expenses:

- Resetting a paid report's payment (or its move) to draft doesn't change the '(payment_)?state' of the report & the related expense
- Reversing the payment (or its move) doesn't change the state of the expense & the expense report BUT sets the payment_state to reverse (and keep the link between move and expense report)

The aim is to backport the expense & expense sheet states behaviour to the one added in saas-16.4 81fe11fabb60de56241ea912a65f41a2d9f8538f

task-3383694

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/149888).

lordkrandel commented:
No time, really, ask someone.

JulienAlardot commented:
@dbkosky ready for review

dbkosky reviewed (commented):
I'm on the fence about the monolithic test. I understand the purpose, and I think it's represents the transitions through the expected series of states nicely. 

The major downside is that it looks difficult to maintain. In any case I think you're going to have to be prepared to argue your case if you want to keep it, since it breaks with established precedent.

JulienAlardot commented:
@dbkosky I moved the test to another file and split it in three parts to make it more readable and easier to work with. Don't know if you think it's better that way

dbkosky reviewed (approved):
looks good to me, I'll forward it to Florian

FlorianGilbert commented on addons/hr_expense/tests/test_expenses_states.py:
```diff
@@ -0,0 +1,313 @@
+
+        # STEP 4: ER with draft move -> Reverse move (Reverts to approve state)
+        self.expense_states_employee_sheet.account_move_id._reverse_moves(
+            default_values_list=[{'invoice_date': fields.Date.today(self.expense_states_employee_sheet)}],
```
Why are we passing the sheet as argument for `today()` function? 
It useless with `today()` but not with `context_today()`.

JulienAlardot replied:
Because I'm an idiot I guess, that's what I wanted to do :facepalm:

FlorianGilbert commented on addons/hr_expense/tests/test_expenses_states.py:
```diff
@@ -0,0 +1,313 @@
+
+        # STEP 4: ER Done & paid -> Reverse move (Change payment state to 'reversed')
+        self.expense_states_company_sheet.account_move_id._reverse_moves(
+            default_values_list=[{'invoice_date': fields.Date.today(self.expense_states_company_sheet)}],
```
Same here

FlorianGilbert commented on addons/hr_expense/views/hr_expense_views.xml:
Changing something in a view in stable, nothing that will break the database but is it really needed?

JulienAlardot replied:
Not really, it will raise anyway

FlorianGilbert commented on addons/hr_expense/models/hr_expense.py:
```diff
@@ -1016,7 +1016,12 @@ def _compute_amount(self):
-            sheet.payment_state = sheet.account_move_id.payment_state or 'not_paid'
+            if not sheet.account_move_id:
+                sheet.payment_state = 'not_paid'
+            elif sheet.account_move_id.reversal_move_id.amount_total == sheet.account_move_id.amount_total:
```
What about using a [float_compare](https://github.com/odoo/odoo/blob/18b2ba6d10cbb466334811f600a0de3905b1cf04/odoo/tools/float_utils.py#L127)?

FlorianGilbert reviewed (changes requested):
Some suggestions
Thank you!

JulienAlardot commented:
@fw-bot up to saas-16.3

FlorianGilbert reviewed (approved):
LGTM 
Thank you :sun_with_face: 

@robodoo r+
